sequence_processing_pipeline/Pipeline.py

```diff
@@ -276,6 +297,69 @@ def _validate_sample_sheet(self, sample_sheet_path):
                               isinstance(x, WarningMessage)]
             return val_sheet
 
+    def _validate_mapping_file(self, mapping_file_path):
+        """
+        Performs validation for mapping-files.
+        :return: If successful, a valid mapping-file. Raises descriptive
+                 PipelineError() on all failures. Warning messages are
+                 appended to self.warnings.
+        """
+        try:
+            df = pd.read_csv(mapping_file_path, delimiter='\t', dtype=str)
+        except pd.errors.ParserError:
+            # ignore parser errors as they obviously prove this is not a
+            # valid mapping file. Return None w/error message
+            raise PipelineError('Cannot parse mapping-file.')
+
+        # first, detect any duplicate column names, regardless of any mixed-
+        # capitalization, and notify the user.
+        d = defaultdict(list)
+        for column in df.columns:
+            d[column.lower()].append(column)
+
+        # generate a list of all unique column names that appear more than
+        # once, regardless of capitalization.
+        duplicate_columns = [col for col in d.keys() if len(d[col]) > 1]
+
+        # generate a list containing lists of duplicate column names in their
+        # original case to report to the user.
+        dupes = [d[column] for column in duplicate_columns]
+
+        if dupes:
+            # column-names are case-insensitive, and must be unique.
+            # return groups of duplicate column names (differentiated only by
+            # a different mixed-case) to the user.
+            raise PipelineError("Mapping-file contains duplicate columns: "
+                                "%s" % ', '.join([str(tpl) for tpl in dupes]))
+
+        # if columns are unique, determine if any columns are missing and/or
+        # unexpected and notify the user.
+        obs = set(df.columns.str.lower())
+        exp = {column.lower() for column in Pipeline.mapping_file_columns}
```

Comment:
Shouldn't this be already lower and a set? Why do we need to lower()?

Reply:
Originally I was going to swap out mapping_file_columns for a list of columns from metapool but I didn't find a good match. This was there just to make sure whatever the source, the values would be lowered. I'll replace it with a comment.